drivers/include/pcf857x.h

```diff
@@ -0,0 +1,557 @@
+    PCF857X_EXP_PCF8574A,   /**< PCF8574A 8 bit I/O expander */
+#endif
+#if IS_USED(MODULE_PCF8575) || DOXYGEN
+    PCF857X_EXP_PCF8575,    /**< PCF8575 16 bit I/O expander */
+#endif
+    PCF857X_EXP_MAX,
+} pcf857x_exp_t;
+
+/**
+ * @brief   PCF857X device initialization parameters
+ */
+typedef struct {
+
+    i2c_t    dev;       /**< I2C device (default I2C_DEV(0)) */
+    uint16_t addr;      /**< I2C slave address offset to the PCF7857X base
+                             address (default 0) */
+    pcf857x_exp_t exp;  /**< PCF857X expander variant used by the device
+                             (default depends on used pseudomodules */
+
+#if IS_USED(MODULE_PCF857X_IRQ) || DOXYGEN
+    gpio_t   int_pin;   /**< MCU interrupt pin or #GPIO_UNDEF if not used (default).
+                             Using interrupt pin has the advantage that inputs
+                             have to be read from expander only if any input
+                             value changes.
+                             @note To use interrupts for expander inputs, this
+                             pin has to be defined. */
+#endif /* MODULE_PCF857X_IRQ */
+} pcf857x_params_t;
+
+#if IS_USED(MODULE_PCF857X_IRQ) || DOXYGEN
+/**
+ * @brief   IRQ event type
+ *
+ * Handling an interrupt of a PCF857x expander requires the driver to access
+ * the device directly via I2C. However, the mutex-based synchronization of
+ * I2C accesses does not work in the interrupt context. Therefore the ISR must
+ * not access the PCF857x expander device directly. Rather, the ISR must only
+ * indicate the occurrence of the interrupt which has to be handled
+ * asynchronously in the thread context.
+ *
+ * The type defines the data structure that is part of each device data
+ * structure to indicate that an interrupt of the device occurred. Since there
+ * is only one interrupt source, only one interrupt can be pending per device.
+ * Thus, only one object of this type per device is required.
+ */
+typedef struct {
+    event_t event;      /**< inherited event data structure */
+    void *dev;          /**< PCF857X device reference */
+} pcf857x_irq_event_t;
+
+#endif /* MODULE_PCF857X_IRQ */
+
+/**
+ * @brief   PCF857X device data structure type
+ */
+typedef struct {
+    pcf857x_params_t params;  /**< device initialization parameters */
+
+    unsigned int pin_num;     /**< number of I/O pins, depends on used expander
+                                   variant */
+    pcf857x_data_t modes;     /**< expander pin modes */
+    pcf857x_data_t in;        /**< expander input pin values */
+    pcf857x_data_t out;       /**< expander output pin values */
+
+#if IS_USED(MODULE_PCF857X_IRQ) || DOXYGEN
+    gpio_isr_ctx_t isr[PCF857X_GPIO_PIN_NUM];  /**< ISR with arg for each expander pin */
+    gpio_flank_t flank[PCF857X_GPIO_PIN_NUM];  /**< interrupt flank for each expander pin */
+    bool enabled[PCF857X_GPIO_PIN_NUM];        /**< enabled flag for each expander pin */
```

Comment:
I guess a bitmap would save a fey bytes here

Reply:
The following tables show the results. For STM32 it requires about 40 byte more ROM to save 12 or 4 bytes of RAM. However, on AVR it requires about 120 or 170 bytes to save 16 or 4 bytes of RAM. I 'm not sure whether this is a good deal.

16-pin expanders
-----------------------
| STM32     |  text   | data  | bss | dec       |
|:--------------|:---------|:-------|:------|:----------|
| array        | 19716 | 540 | 3772 | 24028   |
| bitmap     | 19752 | 540 | 3760 | 24052   |

| AVR     |  text     | data  | bss  | dec      |
|:----------|:----------|:-------|:-------|:---------|
| array    | 18416 | 2250 | 1682 | 22348 |
| bitmap | 18534 | 2250 | 1666 | 22450  |

8-pin expanders
---------------------
| STM32     |  text   | data  | bss | dec       |
|:--------------|:---------|:-------|:------|:----------|
| array        | 19692 | 540 | 3688 | 23920 |
| bitmap     | 19732 | 540 | 3684 | 23956 |

| AVR     |  text     | data  | bss  | dec      |
|:----------|:----------|:-------|:-------|:---------|
| array    | 18114 | 2250 | 1630 | 21994 |
| bitmap | 18282 | 2250 | 1622 | 22154 |

Reply:
Uh I didn't expect that toggling bits would be so expensive, then keep it the way it is